### PR TITLE
Supabase - Storage driver

### LINKED ISSUE #16903: Add native Supabase Storage abstraction
Supabase Storage is S3 'compatible', but has it's own methods and SDKs to manipulate files. Now that #16825 is merged and released, it should be trivial to create a new `@directus/storage-driver-supabase` abstraction.

### PR DESCRIPTION
**Pull request by matt-rolley**
Hey,

I've attempted to add a Supabase storage driver based on the S3 storage driver (where possible). It should help close #16903.

Please let me know if anything needs changing or if I've missed a step (it's my first proper pull request :)) - happy to make any changes that are needed.

Thanks,
Matt

### CODE REVIEW

**Comment by pixelkoduje** [comment]
@matt-rolley Hello, how to run your version to test?  Because it is not known if and when  she will enter the main directus quest.  And I really need it.

**Comment by pixelkoduje** [comment]
If possible, I will be very grateful for releasing as a separate npm package temporarily, e.g. Instead of directus, directus-supa

**Comment by matt-rolley** [comment]
Hey @pixelkoduje,

If you checkout the storage-driver-supabse branch and use the [Running Locally](https://docs.directus.io/contributing/running-locally.html) guide then you should be able to use Supabase storage and database locally. I'd not recommend it though until it's been reviewed/merged into live.

Sorry I can't be more help but let me know if you have any issues using the guide and I'd be happy to try and help :)

**Comment by pixelkoduje** [comment]
What are you waiting for? Until more commits come in and you block adding supabase to directus again? Ehh, merge it with the main project finally. You have already blocked one integration.

**Comment by br41nslug** [comment]
> What are you waiting for? 

[Unfortunately, we can not provide estimates on features or fixes](https://docs.directus.io/getting-started/support.html#can-you-give-an-eta-for-this-feature-fix). As of now, the only way to get an accurate timeline or release date is to [sponsor](https://docs.directus.io/getting-started/support.html#sponsored-work) this task. We recognize that this may be an important feature/fix, but we are a small open-source organization with a lot to triage and complete. If you can't sponsor this ticket, then the next best thing is to increase its priority by giving it a 👍

**Comment by rijkvanzanten** [comment]
> What are you waiting for? Until more commits come in and you block adding supabase to directus again? Ehh, merge it with the main project finally. You have already blocked one integration.

Once a PR is merged, I have to take on all the ongoing maintenance, support, and documentation efforts required to have this feature over time, so I want to make sure it's good to go and up to spec before merging. This is a bigger new feature that needs in-depth reviewing, so it will take a minute. I had no idea this PR was in the works, and can't just drop everything else that was scheduled to do this first 🤷🏻

**Comment by matt-rolley** [comment]
Sorry, I've been away for a week so just catching up properly - appreciate the updates and apologies if I've stepped on any toes, I just thought it'd be an interesting feature to add.

Hoping to contribute more in the future so please let me know if there is a better way for me to help out :)

**Comment by rijkvanzanten** [comment]
> Sorry, I've been away for a week so just catching up properly - appreciate the updates and apologies if I've stepped on any toes, I just thought it'd be an interesting feature to add.

Oh no not at all! Really appreciate this PR and the functionalities it unlocks 😄 My comment earlier was just to explain to @pixelkoduje why running an open source project can't just  "just merge it already" yolo mode all the time 😬 I'm aiming to have it reviewed by this weeks release though!

> Hoping to contribute more in the future so please let me know if there is a better way for me to help out :)

Thanks! PRs are welcome for any Issue (which includes Supabase storage support!) 🙂 Happy to chat [on Discord](https://directus.chat) if you want to talk through specifics.

**Comment by rijkvanzanten** [comment]
@matt-rolley I'm curious to hear your thoughts on resumable uploads in this context. Where does `Upload` from `tus` keep track of the previous parts? Seeing that the Directus API itself doesn't have resumable uploads, does it make sense for this abstraction to have it? Should we instead remove the resumable uploads from the Supabase driver specifically in favor of making it natively supported across the driver abstraction layer

File: packages/storage-driver-supabase/package.json [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	"scripts": {
+		"build": "tsc --project tsconfig.prod.json",
+		"dev": "tsc --watch",
+		"test": "vitest --watch=false"
+	},
+	"dependencies": {
+		"@directus/storage": "workspace:*",
+		"@directus/utils": "workspace:*",
+		"@supabase/storage-js": "^2.5.1",
+		"tus-js-client": "^3.1.0",
```
Comment (rijkvanzanten) [inline packages/storage-driver-supabase/package.json:32]: Lets use fixed versions here

File: packages/storage-driver-supabase/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+	// CHECK: Should src/dest use fullpath?
+	async move(src: string, dest: string) {
+		await this.bucket.move(src, dest);
+	}
+
+	// CHECK: Should src/dest use fullpath?
+	async copy(src: string, dest: string) {
+		await this.bucket.copy(src, dest);
+	}
```
Comment (rijkvanzanten) [inline packages/storage-driver-supabase/src/index.ts:128]: Yes! `root` is configured on the driver level, and injected through `getFullPath`. Without using that, you could break out of the configured root folder.

Comment (matt-rolley) [reply packages/storage-driver-supabase/src/index.ts]: Okay cool, I thought they should use getFullPath, thought I'd added it but guess I missed it. Should be added now, thanks :)

**Comment by matt-rolley** [comment]
> @matt-rolley I'm curious to hear your thoughts on resumable uploads in this context. Where does `Upload` from `tus` keep track of the previous parts? Seeing that the Directus API itself doesn't have resumable uploads, does it make sense for this abstraction to have it? Should we instead remove the resumable uploads from the Supabase driver specifically in favor of making it natively supported across the driver abstraction layer

To be honest I'm not 100% sure how it works, I just saw that [Supabase supported resumable uploads](https://supabase.com/blog/storage-v3-resumable-uploads) so I added it into the storage layer.

I think resumable uploads would be a great feature, especially for the larger files, but if it'd only be supported on the Supabase storage layer then I agree that it makes more sense to add it to the driver abstraction layer. Does it make sense to remove it from this version until it's better tested/added to the driver abstraction layer?

**Comment by rijkvanzanten** [comment]
> Does it make sense to remove it from this version until it's better tested/added to the driver abstraction layer?

Yeah lets do that 👍🏻  Resumable uploads is really something that should be implemented in the Directus API first, and then we can pass that information on to the Supabase (or any other!) driver. I don't think it'll really do anything in it's current state, as Directus will cancel the upload on unexpected errors (rather than retry)

**Comment by matt-rolley** [comment]
Yeah, agreed - sounds like a plan. I've removed it from the codebase now :)

**Review by rijkvanzanten — APPROVED** [review]
This is looking and feeling great @matt-rolley! 🚀 

Really appreciate how closely you've matched the existing code styles and paradigms used. If I didn't know any better, I could've thought this was written by myself 😄

**Comment by dosandco** [comment]
This is super exciting - thanks so much guys!

**Comment by matt-rolley** [comment]
You're welcome :) That's great, really glad it's working nicely!

Looking forward to the release, if any issues pop up, feel free to tag me in them :) happy to help where I can.

Thanks again for the help in getting this live.

**Pull request merged**